### PR TITLE
MOD-8062: Hide internal commands in new ACL category for RediSearch internal commands

### PR DESCRIPTION
**Pull request by raz-mon**
In this PR we introduce the new `_search_internal` ACL command category, to which we register all of our internal commands, i.e., all the `_FT.*` commands. We register these commands ONLY to this command category, such that only the coordinator will be eligible to execute them via the permissions given to its user-defined ACL user.
A correct configuration of the ACL permissions, in which all other users are disallowed to execute RediSearch internal commands by appending the `-@_search_internal` command category to their command permissions.
Notice: These commands change between our OSS and enterprise builds. On enterprise this command category should not be used.

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/module.h [hunk 1]
```diff
@@ -57,6 +57,7 @@ do {                                            \
   RedisModule_ReplyWithStringBuffer(ctx, literal, sizeof(literal) - 1)
 
 #define SEARCH_ACL_CATEGORY "search"
+#define SEARCH_ACL_INTERNAL_CATEGORY "search_internal"
```
Comment (oshadmi) [inline src/module.h:60]: Can we name the internal category with a leading underscore? (as internal commands)

File: src/module.c [hunk 2]
```diff
@@ -988,20 +988,20 @@ int RMCreateSearchCommand(RedisModuleCtx *ctx, const char *name,
 
   int rc = REDISMODULE_OK;
   char *categories;
-  if (!strcmp(aclCategories, "")) {
-    categories = SEARCH_ACL_CATEGORY;
+
+  if (strncmp(name, "_", 1) == 0) {
```
Comment (oshadmi) [inline src/module.c:992]: Consider a function/macro to detect internal commands

File: src/module.c [hunk 3]
```diff
@@ -988,20 +988,20 @@ int RMCreateSearchCommand(RedisModuleCtx *ctx, const char *name,
...
+    // Internal command, register to the internal ACL category
+    rm_asprintf(&categories, strcmp(aclCategories, "") != 0 ? "%s %s" : "%.0s%s", aclCategories, SEARCH_ACL_INTERNAL_CATEGORY);
```
Comment (oshadmi) [inline src/module.c:994]: Should this be done only with oss cluster?

File: src/module.h (hunk 1 above)
Comment (raz-mon) [reply src/module.h:60]: Yes that makes sense 👌

File: src/module.c (hunk 2 above)
Comment (raz-mon) [reply src/module.c:992]: `strncmp` is the detection function 🤔 
A macro is possible but not sure I see a good reason to use one here

File: src/module.c [hunk 4]
```diff
@@ -1069,6 +1069,12 @@ int RediSearch_InitModuleInternal(RedisModuleCtx *ctx, RedisModuleString **argv,
       return REDISMODULE_ERR;
   }
 
+  // Create the `search_internal` ACL command category
+  if (RedisModule_AddACLCategory(ctx, SEARCH_ACL_INTERNAL_CATEGORY) == REDISMODULE_ERR) {
+      RedisModule_Log(ctx, "warning", "Could not add " SEARCH_ACL_INTERNAL_CATEGORY " ACL category, errno: %d\n", errno);
```
[the remark this answers is not in the document]
Comment (raz-mon) [reply src/module.c:1074]: Both.
We log this warning, but also fail loading the module (error)

File: tests/pytests/test_acl.py [hunk 5]
```diff
@@ -8,88 +8,114 @@ def test_acl_category(env):
...
+        # Let's add `write` permissions to `test`
+        conn.execute_command('AUTH', 'default', '')
+        conn.execute_command('ACL', 'SETUSER', 'test', '+@write')
+        conn.execute_command('AUTH', 'test', '123')
+
+        # `test` should now be able to run `write` commands
+        for cmd in WRITE_SEARCH_COMMANDS:
+            try:
+                res = conn.execute_command(cmd)
+                env.assertTrue(False)
```
Comment (oshadmi) [inline tests/pytests/test_acl.py:105]: If should be able to run `write` commands - should reach here? (without an exception)

Comment (oshadmi) [reply tests/pytests/test_acl.py:105]: Oh, there's no `search` category yet?

File: tests/pytests/test_acl.py [hunk 6]
```diff
@@ -8,88 +8,114 @@ def test_acl_category(env):
...
+            except Exception as e:
+                env.assertNotContains("User test has no permissions", str(e))
+
+        # Add `test` search permissions
+        conn.execute_command('AUTH', 'default', '')
+        conn.execute_command('ACL', 'SETUSER', 'test', '+@search', '+@_search_internal')
```
Comment (oshadmi) [inline tests/pytests/test_acl.py:111]: What about failing to run internal commands without `_search_internal` category?

File: tests/pytests/test_acl.py (hunk 5 above)
Comment (raz-mon) [reply tests/pytests/test_acl.py:107]: No, it fails due to an imcomplete command, non-related to permissions.
Added a comment for that.

File: src/module.c [hunk 7]
```diff
@@ -988,18 +989,22 @@ int RMCreateSearchCommand(RedisModuleCtx *ctx, const char *name,
 
   int rc = REDISMODULE_OK;
   char *categories;
-  if (!strcmp(aclCategories, "")) {
-    categories = SEARCH_ACL_CATEGORY;
+  bool is_internal = false;
+
+  if (IS_INTERNAL_COMMAND(name)) {
+    // Internal command, register to the internal ACL category ONLY
+    categories = SEARCH_ACL_INTERNAL_CATEGORY;
```
Comment (oshadmi) [inline src/module.c:996]: Need further discussion - whether _internal_search category is always separated from other categories or should we also register to additional categories?
* If only _internal_search - it needs to be removed explicitly, (usually admin can define `+@all -@slow` etc. which would not remove _internal_search
* If _internal_search is combined with others, `+@slow` could give users permission to search internal commands without intending to

@MeirShpilraien @LiorKogan

**Review by MeirShpilraien — APPROVED** [review]

**Review by oshadmi — APPROVED** [review]

**Pull request merged**